File: util/java/us/ihmc/util/HTTPHelper.java
/*
 * HTTPHelper.java
 *
 * This file is part of the IHMC Util Library
 * Copyright (c) 1993-2016 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

/**
 * HTTPHelper
 *
 * @author      Marco Arguedas      <marguedas@ihmc.us>
 *
 * @version     $Revision: 1.20 $
 *              $Date: 2016/06/09 20:02:46 $
 */

package us.ihmc.util;

import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.StringTokenizer;

import us.ihmc.comm.CommException;
import us.ihmc.comm.CommHelper;
import us.ihmc.comm.ProtocolException;

/**
 *
 */
public class HTTPHelper
{
    public HTTPHelper (CommHelper commHelper)
    {
        _commHelper = Objects.requireNonNull (commHelper, "The CommHelper instance cannot be null");
    }

    public HTTPHelper (Socket socket)
    {
        Objects.requireNonNull (socket, "The socket cannot be null");
        _commHelper = new CommHelper (socket);
    }

    public InputStream getInputStream()
    {
        return _commHelper.getInputStream();
    }

    public OutputStream getOutputStream()
    {
        return _commHelper.getOutputStream();
    }

    /**
     *
     */
    public void sendRequest (String method, String resourceName, Map<HeaderFields, String> params) throws CommException
    {
        String line = method + " " + resourceName + " HTTP/1.1";
        log ("sending request: [" + line + "]");
        _commHelper.sendLine (line);

        if (params != null) {
            for (HeaderFields key : params.keySet()) {
                String value = params.get (key);
                log ("--> " + key + ": " + value);
                _commHelper.sendLine (key + ": " + value);

            }
        }

        _commHelper.sendLine ("");
    }

    public void sendPOSTRequest (String resourceName, Map<HeaderFields, String> params) throws CommException
    {
        this.sendRequest("POST", resourceName, params);
    }

    /**
     *
     */
    public void sendBlankLine() throws CommException
    {
        _commHelper.sendLine("");
    }

    public void sendResponseCode (int responseCode) throws CommException
    {
        _commHelper.sendLine (getResponseCodeMessage (responseCode));
        sendBlankLine();
    }

    String getResponseCodeMessage (int codeNum)
    {
        return "HTTP/1.1 " + getCodeMessage (codeNum);
    }

    private String getCodeMessage (int code)
    {
        String codeMessage;
        int[] shortcut = {0, LEVEL_200, LEVEL_300, LEVEL_400, LEVEL_500, RESPONSE_CODES};
        int i, pos;

        // Below 100 is illegal for HTTP status
        if (code < 100) {
            return http_status_string[LEVEL_500];
        }

        for (i = 0; i < 5; i++) {
            code -= 100;
            if (code < 100) {
                pos = (code + shortcut[i]);
                if (pos < shortcut[i + 1]) {
                    return http_status_string[pos];
                }
                else {
                    // status unknown
                    return http_status_string[LEVEL_500];
                }
            }
        }
        // 600 or above is also illegal

        return http_status_string[LEVEL_500];
    }

    /**
     *
     */
    public void expectResponse (int expectedCode) throws CommException, ProtocolException
    {
        // expect       "HTTP/1.1 <respCode> <respMessage>"
        // followed by: \n\r

        String lineAux = _commHelper.receiveLine();
        StringTokenizer st = new StringTokenizer (lineAux);
        if (st.countTokens() < 2) {
            throw new ProtocolException ("HTTP Protocol Error");
        }

        //ignore the "HTTP/1.1" string
        st.nextToken();

        int respCode = Integer.MAX_VALUE;
        try {
            respCode = Integer.parseInt(st.nextToken());
        }
        catch (NumberFormatException ex) {
            throw new ProtocolException ("HTTP Protocol Error " + ex.getMessage());
        }

        if (respCode != expectedCode) {
            String errMessage = respCode + " ";
            while (st.hasMoreTokens()) {
                errMessage += st.nextToken();
                if (st.hasMoreTokens()) {
                    errMessage += " ";
                }
            }

            throw new ProtocolException ("Expected " + expectedCode + ". Error occurred: \"" + errMessage + "\"");
        }
        
        lineAux = _commHelper.receiveLine();
        if (!"".equals(lineAux)) {
            throw new ProtocolException ("Expected blank line. Obtained: [" + lineAux + "]");
        }
    }

    public int getExpectedResponseCode() throws CommException, ProtocolException
    {
        // expect       "HTTP/1.1 <respCode> <respMessage>"
        // followed by: \n\r

        String lineAux = _commHelper.receiveLine();
        StringTokenizer st = new StringTokenizer (lineAux);
        if (st.countTokens() < 2) {
            throw new ProtocolException ("HTTP Protocol Error");
        }

        //ignore the "HTTP/1.1" string
        st.nextToken();

        int respCode = Integer.MAX_VALUE;
        try {
            respCode = Integer.parseInt (st.nextToken());
        }
        catch (NumberFormatException ex) {
            throw new ProtocolException ("HTTP Protocol Error " + ex.getMessage());
        }

        lineAux = _commHelper.receiveLine();
        if (!lineAux.isEmpty()) {
            throw new ProtocolException ("Expected blank line. Obtained: [" + lineAux + "]");
        }

        return respCode;
    }

    /**
     *
     */
    public Map<HeaderFields, String> readHeaderFields() throws CommException
    {
        String lineAux;
        Map<HeaderFields, String> headerFields = new HashMap<HeaderFields, String>();

        while ( (lineAux = _commHelper.receiveLine()) != null ) {
            if (lineAux.equals ("")) {
                // end of header.
                break;
            }

            HeaderFields key;
            String value;
            if (lineAux.startsWith (HeaderFields.RequestMethod.name())) {
                StringTokenizer st = new StringTokenizer (lineAux, " ");
                log ("received " + lineAux);
                key = HeaderFields.RequestMethod;
                value = st.nextToken();
                headerFields.put (key, value);
                key = HeaderFields.ResourceName;
                value = st.nextToken();
                headerFields.put (key, value);
                key = HeaderFields.Version;
                value = st.nextToken().replace ("HTTP/", "");
                headerFields.put (key, value);
            }
            else {
                StringTokenizer st = new StringTokenizer (lineAux, ":");
                key = HeaderFields.valueOf (st.nextToken().trim());
                value = st.nextToken().trim();
                log ("received " + key + ": " + value);
                headerFields.put (key, value);
            }
        }

        return headerFields;
    }

    public byte[] receiveContent (int contentLength) throws ProtocolException, CommException
    {
        if (contentLength < 0) {
            throw new ProtocolException ("The content length is negative");
        }

        return _commHelper.receiveBlob (contentLength);
    }

    public void sendField (HTTPHelper.HeaderFields name, String value) throws CommException, ProtocolException
    {
        if ((name == null) || (value == null)) {
            throw new ProtocolException ("Null parameters");
        }

        _commHelper.sendLine (name + ": " + value);
    }

    public void sendContent (String result) throws ProtocolException, CommException
    {
        if (result == null) {
            throw new ProtocolException ("Null results");
        }

        _commHelper.sendLine (result);
    }

    public void sendContent (byte[] buf, int length) throws ProtocolException, CommException
    {
        if (buf == null) {
            throw new ProtocolException ("Null buffer");
        }

        _commHelper.sendBlob (buf, 0, length);
    }

    public void setExpectedResponseCode (int code)
    {
        _expectCode = true;
        _expectedHTTPCode = code;
    }

    private void log (String msg)
    {
        if (DEBUG) {
            System.out.println("[HTTPHelper] " + msg);
        }
    }


    private final boolean DEBUG = false;
    private final CommHelper _commHelper;
    private boolean _expectCode = false;
    private int _expectedHTTPCode;

    public enum HeaderFields
    {
        Version,
        RequestMethod,
        ResourceName,
        Action,
        ActionType,
        ContentType,
        ContentLength,
        RequestorUUID,
        Expect,
        NodeUUID,
        InstanceUUID,
        SrcNodeUUID,
        DestNodeUIID,
        ServiceName,
        MethodName,
        MethodSignature,
        ResInfoQueryType,
        Asynchronous,
        VMContainer
    }

    private static final String[] http_status_string = {
            "100 Continue",
            "101 Switching Protocols",
            "102 Processing",
            "200 OK",
            "201 Created",
            "202 Accepted",
            "203 Non-Authoritative Information",
            "204 No Content",
            "205 Reset Content",
            "206 Partial Content",
            "207 Multi-Status",
            "300 Multiple Choices",
            "301 Moved Permanently",
            "302 Found",
            "303 See Other",
            "304 Not Modified",
            "305 Use Proxy",
            "306 unused",
            "307 Temporary Redirect",
            "400 Bad Request",
            "401 Authorization Required",
            "402 Payment Required",
            "403 Forbidden",
            "404 Not Found",
            "405 Method Not Allowed",
            "406 Not Acceptable",
            "407 Proxy Authentication Required",
            "408 Request Time-out",
            "409 Conflict",
            "410 Gone",
            "411 Length Required",
            "412 Precondition Failed",
            "413 Request Entity Too Large",
            "414 Request-URI Too Large",
            "415 Unsupported Media Type",
            "416 Requested Range Not Satisfiable",
            "417 Expectation Failed",
            "418 unused",
            "419 unused",
            "420 unused",
            "421 unused",
            "422 Unprocessable Entity",
            "423 Locked",
            "424 Failed Dependency",
            "425 No code",
            "426 Upgrade Required",
            "500 Internal Server Error",
            "501 Method Not Implemented",
            "502 Bad Gateway",
            "503 Service Temporarily Unavailable",
            "504 Gateway Time-out",
            "505 HTTP Version Not Supported",
            "506 Variant Also Negotiates",
            "507 Insufficient Storage",
            "508 unused",
            "509 unused"
    };

    private static final int LEVEL_200 = 3;
    private static final int LEVEL_300 = 11;
    private static final int LEVEL_400 = 19;
    private static final int LEVEL_500 = 46;
    private static final int RESPONSE_CODES = 56;
}

File: util/cpp/HTTPHelper.cpp
/*
 * HTTPHelper.cpp
 *
 * This file is part of the IHMC Util Library
 * Copyright (c) 1993-2016 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

#include "HTTPHelper.h"

#include "CommHelper2.h"
#include "StringStringHashtable.h"

#define BUFF_LENGTH 1024

using namespace NOMADSUtil;

HTTPHelper::HTTPHelper (void)
{
    _bExpectCode = false;
    _ui16ExpectedHTTPCode = 0;
    _pContent = NULL;
    _ui32ContentLength = 0;
    _ui16LastResponseCode = 0;
    _bDeleteCommHelper = false;

    _requestMethod = NULL;
    _resourceName = NULL;

    _pHeaderFieldsHT = new StringStringHashtable();
}

HTTPHelper::~HTTPHelper (void)
{
    if (_bDeleteCommHelper && _pch) {
        delete _pch;
    }
    
    delete _pHeaderFieldsHT;
}

int HTTPHelper::init (Socket *pSock)
{
    _pch = new CommHelper2();
    _pch->init (pSock);
    _bDeleteCommHelper = true;
    return 0;
}

int HTTPHelper::init (CommHelper2 *pch)
{
    _pch = pch;
    return 0;
}

int HTTPHelper::init (Reader *pReader, Writer *pWriter)
{
    _pch = new CommHelper2();
    _pch->init (pReader, pWriter);
    _bDeleteCommHelper = true;
    return 0;
}

int HTTPHelper::sendField (const char* pszFieldName, const char* pszFieldValue)
{
    char buf[BUFF_LENGTH];
    strcpy (buf, pszFieldName);
    strcat (buf, ": ");
    strcat (buf, pszFieldValue);

    try {
        _pch->sendLine (buf);
    }
    catch (Exception) {
        return -1;
    }

    return 0;
}

int HTTPHelper::sendHeader (const char *pszMethodName, const char * pszResourceName, StringStringHashtable &headerFields)
{
    char buf[BUFF_LENGTH];
    strcpy (buf, pszMethodName);
    strcat (buf, " ");
    strcat (buf, pszResourceName);
    strcat (buf, " HTTP/1.1");

    try {
        _pch->sendLine (buf);
    
        for (StringStringHashtable::Iterator i = headerFields.getAllElements(); !i.end(); i.nextElement()) {
            strcpy (buf, i.getKey());
            strcat (buf, ": ");
            strcat (buf, i.getValue());
            _pch->sendLine (buf);
        }
        if (_bExpectCode) {
            strcpy (buf, "Expect: ");
            strcat (buf, getCodeMessage (_ui16ExpectedHTTPCode));
            _pch->sendLine (buf);
        }
        sendBlankLine();
    }
    catch (CommException) {
        return -1;
    }

    return 0;
}

int HTTPHelper::expectResponseCode (uint16 ui16ResponseCode)
{
    char buf[BUFF_LENGTH];
    
    try {
        _pch->receiveLine (buf, BUFF_LENGTH);
    }
    catch (CommException) {
        return -1;
    }
    
    char *pszAux, *pszCode;
    strtok_mt (buf, " ", &pszAux);
    pszCode = strtok_mt (NULL, " ", &pszAux);

    itoa (buf, ui16ResponseCode);
    if (strcmp (pszCode, buf) != 0) {
        _ui16LastResponseCode = atoi (pszCode);
        return -2;
    }

    try {
        int rc = _pch->receiveLine (buf, BUFF_LENGTH);
        if (strcmp ("", buf) != 0) {
            return -3;
        }
    }
    catch (CommException) {
        return -4;
    }

    return 0;
}

int HTTPHelper::checkAndSendResponseCode (uint16 ui16ResponseCode)
{
    if (_bExpectCode && (ui16ResponseCode == _ui16ExpectedHTTPCode)) {
        return sendResponseCode (_ui16ExpectedHTTPCode); //usually 100
    }
    return -1; // no response code was expected
}

uint16 HTTPHelper::getExpectedResponseCode (void)
{
    return _ui16ExpectedHTTPCode; // = 0 if no response code is expected
}

int HTTPHelper::setExpectedResponseCode (uint16 ui16ResponseCode)
{
    _bExpectCode = true;
    _ui16ExpectedHTTPCode = ui16ResponseCode;
    _pHeaderFieldsHT->put ("Expect", (char*) getCodeMessage(ui16ResponseCode));
    return 0;
}

int HTTPHelper::sendContent (void *pContent, uint32 ui32ContentLength)
{
    try {
        _pch->sendBlob (pContent, ui32ContentLength);
    }
    catch (CommException) {
        return -1;
    }
    return 0;
}

int HTTPHelper::getLastResponseCode()
{
    return _ui16LastResponseCode;
}

int HTTPHelper::receiveHeader (const char* pszProtoLine)
{
    char * pszTmp;

    _requestMethod = strtok_mt (pszProtoLine, " ", &pszTmp);
    _resourceName  = strtok_mt (NULL, " ", &pszTmp);
    
    return receiveHeader();
}

int HTTPHelper::receiveHeader()
{
    char buf[BUFF_LENGTH];
    _pHeaderFieldsHT->removeAll();
    _bExpectCode = false;

    try {
        while (true) {
            _pch->receiveLine (buf, BUFF_LENGTH);
            //printf ("HTTPHelper:: line rcvd: [%s]\n", buf);
            if (strcmp (buf, "") == 0) {
                //end of header
                break;
            }
            else {
                // parse the line and add it to the hashtable
                char *pszKey, *pszValue, *pszAux;
                pszKey = strtok_mt (buf, " :", &pszAux);
                pszValue = strtok_mt (NULL, " :", &pszAux);
                if (strcmp (pszKey, "Content-Length") == 0) {
                    _ui32ContentLength = atoi (pszValue);
                }
                else if (strcmp (pszKey, "Expect") == 0) {
                    pszValue = strtok_mt (pszValue, "- ", &pszAux);
                    _ui16ExpectedHTTPCode = atoi (pszValue);
                    _bExpectCode = true;
                }
                _pHeaderFieldsHT->put (pszKey, pszValue);
            }
        }
    }
    catch (CommException) {
        return -1;
    }
    return 0;
}

bool HTTPHelper::hasHeaderFieldValue (const char *pszKey)
{
    return (_pHeaderFieldsHT->get (pszKey) != NULL);
}

String HTTPHelper::getHeaderFieldValue (const char *pszKey)
{
    return _pHeaderFieldsHT->get (pszKey);
}

String HTTPHelper::getResourceName()
{
    return _resourceName;
}

String HTTPHelper::getRequestMethod()
{
    return _requestMethod;
}

void * HTTPHelper::receiveContent (void)
{
    if (_ui32ContentLength > 0) {
        _pContent = malloc (_ui32ContentLength);
        try {
            _pch->receiveBlob (_pContent, _ui32ContentLength);
        }
        catch (CommException) {
            return NULL;
        }
        return _pContent;
    }
    return NULL;
}

uint32 HTTPHelper::getContentLength()
{
    return _ui32ContentLength;
}

int HTTPHelper::receiveContent (void *pContent, uint32 ui32ContentLength)
{
    if (_ui32ContentLength > 0) {
        try {
            _pch->receiveBlob (pContent, ui32ContentLength);
        }
        catch (CommException) {
            return -2;
        }
        return 0;
    }
    return -1;
}

int HTTPHelper::sendResponseCode (uint16 ui16ResponseCode)
{
    try {
        _pch->sendLine (getResponseCodeMessage (ui16ResponseCode));
        _pch->sendLine ("");
    }
    catch (CommException) {
        return -1;
    }
    return 0;
}

String HTTPHelper::getResponseCodeMessage (uint16 ui16CodeNum)
{
    String s = "HTTP/1.1 ";
    s += getCodeMessage (ui16CodeNum);
    return s;
}

int HTTPHelper::sendBlankLine (void)
{
    try {
        _pch->sendLine("");
    }
    catch (CommException) {
        return -1;
    }
    return 0;
}

int HTTPHelper::flush()
{
    try {
        return _pch->getWriterRef()->flush();
    }
    catch (CommException) {
        return -1;
    }
}

/////// Private Methods ///////

const char * HTTPHelper::getCodeMessage (uint16 ui16Code)
{
    String codeMessage;
    static int shortcut[6] = {0, LEVEL_200, LEVEL_300, LEVEL_400, LEVEL_500, RESPONSE_CODES};
    int i, pos;

    // Below 100 is illegal for HTTP status
    if (ui16Code < 100) {
        return http_status_string[LEVEL_500];
    }

    for (i = 0; i < 5; i++) {
        ui16Code -= 100;
        if (ui16Code < 100) {
            pos = (ui16Code + shortcut[i]);
            if (pos < shortcut[i + 1]) {
                return http_status_string[pos];
            }
            else {
                // status unknown
                return http_status_string[LEVEL_500];
            }
        }
    }
    // 600 or above is also illegal

    return http_status_string[LEVEL_500];
}

